### PR TITLE
Preserve scroll offset when switching rooms

### PR DESCRIPTION
**Pull request by richvdh**
When we change rooms, save the scroll offset, and restore the scroll when we
switch back.

Hopefully this fixes https://github.com/vector-im/vector-web/issues/80.

### CODE REVIEW

File: src/components/structures/RoomView.js [hunk 1]
```diff
@@ -803,6 +803,96 @@ module.exports = React.createClass({
         scrollNode.scrollTop = scrollNode.scrollHeight;
     },
 
+    // scroll the event view to put the given event at the bottom.
+    //
+    // pixel_offset gives the number of pixels between the bottom of the event
+    // and the bottom of the container.
+    scrollToEvent: function(event_id, pixel_offset) {
```
Comment (ara4n) [inline src/components/structures/RoomView.js]: please can we use camelCase in JS?  (there's some older snake_case stuff hanging around, but it should be defunct)

File: src/components/structures/MatrixChat.js [hunk 2]
```diff
@@ -590,6 +608,7 @@ module.exports = React.createClass({
                 case this.PageTypes.RoomView:
                     page_element = (
                         <RoomView
+                            ref="roomview"
```
Comment (ara4n) [inline src/components/structures/MatrixChat.js]: suggest roomView for camelcasing

File: src/components/structures/RoomView.js [hunk 3]
```diff
@@ -803,6 +803,96 @@ module.exports = React.createClass({
...
+            // looking for a non-existent event.
+            //
+            // for now, just scroll to the top of the buffer.
+            console.log("Refusing to scroll to unknown event "+event_id);
+            scrollNode.scrollTop = 0;
+            return;
+        }
+
+        // we might need to roll back the messagecap (to generate tiles for
+        // older messages). Don't roll it back past the timeline we have, though.
```
Comment (ara4n) [inline src/components/structures/RoomView.js]: might be worth clarifying this to explain that this doesn't involve backfilling - it's just generating more tiles for events we already have in our timeline. i got a bit stuck reading the code before I realised what was going on.

File: src/components/structures/RoomView.js [hunk 4]
```diff
@@ -803,6 +803,96 @@ module.exports = React.createClass({
...
+    // get the current scroll position of the room, so that it can be
+    // restored when we switch back to it
+    getScrollState: function() {
+        // we don't save the absolute scroll offset, because that
+        // would be affected by window width, zoom level, amount of scrollback,
+        // etc.
+        //
+        // instead we save the id of the last fully-visible event, and the
+        // number of pixels the window was scrolled below it - which will
+        // hopefully be near enough.
```
Comment (ara4n) [inline src/components/structures/RoomView.js]: this is very nice :)

**Comment by ara4n** [comment]
lgtm other than stupid cosmetic niggles above, and the concern that it's only going to actually be useful if it also applies a scroll lock when you're scrolled up.  However, i vaguely recall that this once worked - /me pings @dbkr to find out.

File: src/components/structures/RoomView.js (hunk 1 above)
Comment (richvdh) [inline src/components/structures/RoomView.js]: done

File: src/components/structures/MatrixChat.js (hunk 2 above)
Comment (richvdh) [inline src/components/structures/MatrixChat.js]: done

File: src/components/structures/RoomView.js (hunk 3 above)
Comment (richvdh) [inline src/components/structures/RoomView.js]: done.

**Comment by richvdh** [comment]
hopefully this now preserves the scroll state on backfill too.

**Comment by ara4n** [comment]
This looks great to me, modulo just wanting to understand the question above - huge thanks.

Am I right in thinking that it is also fixing https://github.com/vector-im/vector-web/issues/287? Specifically, that as you scroll towards the top of the page, it no longer loses your scroll offset as it backfills?  In practice it certainly seems improved, although there's still some kind of nasty scroll jump if you literally hit the top of the page when triggering a backfill.  Do you expect that to be fixed?

**Comment by richvdh** [comment]
Yeah, it's supposed to maintain your position as backfill happens. I'm certainly prepared to believe there are still issues, because it's quite a complicated dance, but it seemed to work for me. Can you try to explain the jump you're seeing?

**Comment by ara4n** [comment]
If you hammer hard on page-up or drag the scrollbar repatedly against the top such that it backfills twice in a row, it seems like the second backfill causes the scroll offset to jump rather than the contents being prepended without a jump.

This *might* just be a race though - that the first pageup triggers a backfill that completes successfully, and doesn't change the scroll lock.  The second pageup then causes a legitimate jump, which could be mistaken for the scroll lock failing. So this may be just me being a crank - wdyt?

**Comment by ara4n** [comment]
i'm going to merge this anyway as i'm in danger of accidentally fixing bugs it's already fixed (or at least conflicting more badly with it)

**Comment by richvdh** [comment]
> If you hammer hard on page-up or drag the scrollbar repatedly against the top such that it backfills twice in a row, it seems like the second backfill causes the scroll offset to jump rather than the contents being prepended without a jump.

It strikes me that these are two different cases. If you're dragging the scrollbar, you are effectively keeping the scroll at the top so are forcing the contents to be prepended without a jump. OTOH, if you're hammering hard on pageup, the scroll bar does get a chance to move down. In short: I'd *expect* this behaviour if you are dragging the scrollbar, but not if you are hitting pageup.

> This *might* just be a race though - that the first pageup triggers a backfill that completes successfully, and doesn't change the scroll lock.  The second pageup then causes a legitimate jump, which could be mistaken for the scroll lock failing. So this may be just me being a crank - wdyt?

I don't quite understand what you mean. There *shouldn't* be an opportunity for a race here. Hitting the top of the window should be ignored while there is a backfill in progress.

It's hard to say if there is or isn't an issue here; it seems to work fine if you use it slowly, but if you use it quickly it's hard to keep track of what you ought to be seeing.

It might be worth observing that if you have a tall window, such that one backfill isn't enough, you can see the scrollbar update twice as the two backfills happen, while the events stay put. Which suggests that it's more-or-less doing the right thing on a double-backfill?

**Pull request merged**